tcp/tcp.c

```diff
@@ -0,0 +1,382 @@
+// Copyright 2016 by BeeGee, see LICENSE.txt
+//
+// Adapted from: github.com/tuanpmt/esp_bridge, Created on: Mar 4, 2015, Author: Minh
+// Adapted from: rest.c, Author: Thorsten von Eicken
+
+#include "esp8266.h"
+#include "c_types.h"
+#include "ip_addr.h"
+#include "tcp.h"
+#include "cmd.h"
+
+#define TCP_DBG
+
+#ifdef TCP_DBG
+#define DBG_TCP(format, ...) os_printf(format, ## __VA_ARGS__)
+#else
+#define DBG_TCP(format, ...) do { } while(0)
+#endif
+
+typedef struct {
+	char			*host;
+	uint32_t		port;
+	ip_addr_t		ip;
+	struct espconn	*pCon;
+	char			*data;
+	uint16_t		data_len;
+	uint16_t		data_sent;
+	uint32_t		resp_cb;
+	uint8_t			conn_num;
+	uint8_t			sock_mode;
+} TcpClient;
+
+
+// Connection pool for TCP socket clients. Attached MCU's just call TCP_setup and this allocates
+// a connection, They never call any 'free' and given that the attached MCU could restart at
+// any time, we cannot really rely on the attached MCU to call 'free' ever, so better do without.
+// Instead, we allocate a fixed pool of connections an round-robin. What this means is that the
+// attached MCU should really use at most as many TCP connections as there are slots in the pool.
+#define MAX_TCP 4
+static TcpClient tcpClient[MAX_TCP];
```

Comment:
Put together TCP and UDP socket functions into one class, therefor the memory consumption is limited. Otherwise left as it is.
I guess it was fixed memory allocation, because the sockets opened by the uC is assigned from the pool in a round robin as there is no guarantee that the uC closes the socket (might be because of a reset).